### PR TITLE
Add support for `save_tree_error` train argument to preserve `record_evals`

### PR DESCRIPTION
**Pull request by jeancochrane**
This PR updates `train_lightgbm` to add support for a boolean `save_tree_error` argument, which allows the caller to specify that the model should use its training data as a validation set in order to save a `record_evals` attribute that records the error for each tree in the model. The PR also tweaks `lgbm_save` and `lgbm_load` in order to support this attribute being (de)serialized when saving or loading the model.

Connects https://github.com/ccao-data/model-res-avm/pull/106.

### CODE REVIEW

File: R/lightgbm.R [hunk 1]
```diff
@@ -135,6 +135,7 @@ train_lightgbm <- function(x,
                            feature_pre_filter = FALSE,
                            free_raw_data = TRUE,
                            verbose = 0,
+                           save_tree_error = FALSE,
```
Comment (jeancochrane) [inline R/lightgbm.R:143]: I don't love that this argument doesn't match [the `lgbm.train` interface](https://lightgbm.readthedocs.io/en/stable/R/reference/lgb.train.html); previously I had tried adding a `valids` keyword argument for better alignment, but I thought that was too confusing alongside the existing `validation` keyword arg, and plus requiring the caller to format the valids set for record evals leaves more room for error in creating the `Dataset` object. This approach seemed like a reasonable middle ground but I'm open to alternatives!

File: R/lightgbm.R [hunk 2]
```diff
@@ -238,6 +239,18 @@ train_lightgbm <- function(x,
     free_raw_data = free_raw_data
   )
 
+  if (save_tree_error) {
+    print("Enabling save_tree_error")
+    if (exists("valids")) {
+      print("valids exists; adding tree_errors to it")
+      valids$tree_errors <- d
+    } else {
```
Comment (jeancochrane) [inline R/lightgbm.R]: This conditional branch ensures that the valids set used for record evals doesn't accidentally clobber the valids set created for validation above.

File: R/lightgbm.R [hunk 3]
```diff
@@ -247,11 +260,15 @@ train_lightgbm <- function(x,
...
-  if (!is.null(early_stop) && validation > 0) {
+  if (exists("valids")) {
     main_args$valids <- quote(valids)
+  }
+  if (!is.null(early_stop) && validation > 0) {
     main_args$early_stopping_rounds <- early_stop
   }
 
+  print("Calling lgb.train with the following args:")
+  print(main_args)
```
Comment (jeancochrane) [inline R/lightgbm.R:271]: I found this to be useful as I was debugging training behavior, but I'm happy to remove it if you don't think it's more widely useful.

File: R/lightgbm.R [hunk 4]
```diff
@@ -238,6 +242,18 @@ train_lightgbm <- function(x,
     free_raw_data = free_raw_data
   )
 
+  if (save_tree_error) {
+    print("Enabling save_tree_error")
```
Comment (jeancochrane) [inline R/lightgbm.R]: I found these print statements to be useful as I was debugging training behavior, but I'm happy to remove it if you're worried about cluttering up the logs. Or perhaps we only print it if `verbose = TRUE`? I'm still not sure what the idiomatic way to do log levels is in R.

File: R/lightgbm.R (hunk 1 above)
Comment (dfsnow) [reply R/lightgbm.R:143]: I like it! It's our shim package specifically for our pipeline. I think it's fine to add what we need as long as _we_ understand the interface.

File: R/lightgbm.R (hunk 4 above)
Comment (dfsnow) [reply R/lightgbm.R]: **suggestion:** I would do prints only if `verbose > 0` and replace the `print()` calls with `message()`.

**Review by dfsnow — APPROVED** [review]
@jeancochrane This looks good to me other than the minor changes around verbosity and printing debug statements. It does look like we need to be careful with this during CV though. Per the `lightgbm` docs:

> **early_stopping_rounds** - Activates early stopping. When this parameter is non-null, training will stop if the evaluation of any metric on any validation set fails to improve for early_stopping_rounds consecutive boosting rounds. If training stops early, the returned model will have attribute best_iter set to the iteration number of the best iteration.

So we'll need to be sure to set it _only_ for the final model, not during CV.

File: R/lightgbm.R (hunk 4 above)
Comment (jeancochrane) [reply R/lightgbm.R]: Good call, done in [8940acc](https://github.com/ccao-data/lightsnip/pull/10/commits/8940acc93fc1bdd5afb23210ae93789adbb25613)!

**Comment by jeancochrane** [comment]
> This looks good to me other than the minor changes around verbosity and printing debug statements. It does look like we need to be careful with this during CV though. Per the lightgbm docs: [...] So we'll need to be sure to set it only for the final model, not during CV.

Good catch @dfsnow! To formalize this decision, I added a line that raises an error if `save_tree_error` is `TRUE` and `validation > 0` (in addition to updating the docs). If that seems like a reasonable choice to you, I can go ahead and merge this.

**Comment by dfsnow** [comment]
> > This looks good to me other than the minor changes around verbosity and printing debug statements. It does look like we need to be careful with this during CV though. Per the lightgbm docs: [...] So we'll need to be sure to set it only for the final model, not during CV.
> 
> Good catch @dfsnow! To formalize this decision, I added a line that raises an error if `save_tree_error` is `TRUE` and `validation > 0` (in addition to updating the docs). If that seems like a reasonable choice to you, I can go ahead and merge this.

Good call. Let's merge it!
